applications/TrilinosApplication/custom_strategies/convergencecriterias/trilinos_mixed_generic_criteria.h

```diff
@@ -0,0 +1,376 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Jordi Cotela, Riccardo Rossi, Carlos Roig and Ruben Zorrilla
+//
+
+#ifndef KRATOS_TRILINOS_MIXED_GENERIC_CRITERIA_H
+#define	KRATOS_TRILINOS_MIXED_GENERIC_CRITERIA_H
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "includes/define.h"
+#include "includes/model_part.h"
+#include "solving_strategies/convergencecriterias/convergence_criteria.h"
+
+// Application includes
+
+namespace Kratos
+{
+///@addtogroup TrilinosApplication
+///@{
+
+///@name Kratos Classes
+///@{
+
+/// Convergence criteria for mixed vector-scalar problems.
+/**
+ This class implements a convergence control based on a nodal vector variable and
+ a nodal scalar variable. The error is evaluated separately for each of them, and
+ relative and absolute tolerances for both must be specified.
+ */
+template< class TSparseSpace, class TDenseSpace >
+class TrilinosMixedGenericCriteria : public ConvergenceCriteria< TSparseSpace, TDenseSpace >
+{
+public:
+    ///@name Type Definitions
+    ///@{
+
+    KRATOS_CLASS_POINTER_DEFINITION(TrilinosMixedGenericCriteria);
+
+    typedef ConvergenceCriteria< TSparseSpace, TDenseSpace > BaseType;
+
+    typedef typename BaseType::TDataType TDataType;
+
+    typedef typename BaseType::DofsArrayType DofsArrayType;
+
+    typedef typename BaseType::TSystemMatrixType TSystemMatrixType;
+
+    typedef typename BaseType::TSystemVectorType TSystemVectorType;
+
+    typedef typename DofUpdater<TSparseSpace >::UniquePointer DofUpdaterPointerType;
+
+    typedef std::vector<std::tuple<VariableData, TDataType, TDataType>> ConvergenceVariableListType;
+
+    typedef std::size_t KeyType;
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Constructor.
+
+    /**
+     * @brief Construct a new Trilinos Mixed Generic Criteria object
+     *
+     * @param rConvergenceVariablesList List containing tuples with the convergence variables to be checked. The tuples are set as <Variable, relative tolerance, absolute tolerance>
+     */
+    TrilinosMixedGenericCriteria(const ConvergenceVariableListType& rConvergenceVariablesList)
+        : ConvergenceCriteria<TSparseSpace, TDenseSpace>()
+        , mVariableSize([&] (const ConvergenceVariableListType& rList) -> int {return rList.size();} (rConvergenceVariablesList))
+        , mVariableDataVector([&] (const ConvergenceVariableListType& rList) -> std::vector<std::shared_ptr<VariableData>> {
```

Comment:
i think that this

~~~
std::vector<std::shared_ptr<VariableData>>
~~~

is really incorrect.

you cannot create a shared_ptr to an existing object

Reply:
I am quite certain I have a better solution for this but no time to post it atm

Reply:
@RiccardoRossi suggests to create an `std::vector<VariableData*>`. @roigcarlo after thinking a bit about it, isn't this what we already tried?

Reply:
> I am quite certain I have a better solution for this but no time to post it atm

Can you briefly comment it? (just the idea)

Reply:
now I have a moment to outline the basic idea:
I worked on basically the same some time ago but I didn't yet manage to finish it yet: https://github.com/KratosMultiphysics/Kratos/compare/experimental/unify-conv-criteria

The idea is the same as we use for many processes: The variables are passed in by name and then using the `KratosComponents` they are read in. This way you won't have this problem that you are having. Here is how I did it: https://github.com/KratosMultiphysics/Kratos/compare/experimental/unify-conv-criteria#diff-8cc89e329e137930d074d05e39603cbeR124-R207

Here is an example of a process that uses what I explained: https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/processes/integration_values_extrapolation_to_nodes_process.cpp#L51-L64

Using it like this will also be necessary for #3185, so I think it makes sense to do it like this right away

Reply:
I've been speaking with @RiccardoRossi . What he has in mind is a little differetn. We can create a `std::vector<VariableData*>` and also change the `typedef std::vector<std::tuple<VariableData*, TDataType, TDataType>> ConvergenceVariableListType`. 

If that works we can obtain the pointer from the tuple without using &

Reply:
but still how would you obtain this from a `string` (which is required for #3185)?

Reply:
Now I am lost. From a string? in this PR at least we are passing the variable data, not the name of the variable :S

Reply:
Ok I saw your pr. The idea is to avoid doing this: ```if (KratosComponents<DoubleVariableType>::Has(variable_name))```

So you obtain directly the variable without checking the list of all double variables and all array variables (in the worst case)

But technically what you did would work here as well.

Reply:
I'd like to add that what @philbucher would do the job in this case. However, I'd insist a bit more in fixing this approach. I really like the Python input `[(VARIABLE, tol_1, tol_2)]. Using the `KratosComponents` implies `[(VARIABLE.Name(), tol_1, tol_2)]`.

Reply:
We can have another (python) constructor 

What I am also telling is that we NEED a Parameters based constructor for #3185

Reply:
I didn't know the full extend of #3185 and now I have a better picture of what is happening.
I agree that we need that constructor too.

Now my concerns (this is not any kind of criticism, it's just I did not see this in the other discussion and I want to state it somewhere):

Although I see the point on accessing components directly inside the criteria provided that #3185 is a thing, I would have still prefer the option of not having a parameters constructor.

To me (and I want to make very clear that this is an opinion, I have no much experience in this field) the decision of having a parameters constructor is only driven by the fact that we are basing everything in a generic factory patter than can create virtually every convergence criteria, strategy, or by that matter class, in Kratos. This is not necessary bad, and for what I can see is the preferred criteria. But this enforces a single unique signature for every class constructor.

If we were using, for instance, an specialized factory for converge criteria that know how to handle all criteria in Kratos, we could free from the parameters restriction and handle them in the factory itself.

I.e: (very pseudocode)
```C++
class TrilinosConvergeCriteriaFactory : ConvergenceCriteriaFactory {
public:
    ConvergeCriteria * Create(std::string name, Parameters params) {
        if (name == "and_criteria") {...}
        if (name == "or_criteria") {...}
        if (name == "generic") {
                // You extract the variabels from the comps here
                var1, var2  = GetVariables(params)

                // You know which is the consttructor
                return GetComponent<ConvergeCriteria>::Get(name).Create(.....) 
        }
    } 
}
```

As you can the criteria here does not need to know about `KratosComponents`, which is a very internal thing to me, and it is transparent to the type of the vars. You could perfectly use this class with a array_6 variable ( I ignore if this has any physical relevance) without having to add extra if/checks inside the cc itself, but rather in the factory, which should be the proper place in my opinion, or directly from python.

That will also for example would avoid having to jump back and forth from the factory to the converge criteria when creating `and` and `or` criteria (that's another discussion...)

Again probably this is a cross discussion that also affects #3185 and I am still reading all comments and responses. Maybe we could discuss about this in the next technical committee.

Reply:
I found this is important but out of the scope of the PR. I suggest merging this as it would solve the problems that the CFD elements without nodal properties have when checking the pressure convergence with the current mixed criteria. We can continue this general discussion and implement the parameters-based constructor later on.